crates/agglayer-node/src/rate_limiting/local/inner/component.rs

```diff
@@ -0,0 +1,46 @@
+use super::{limiter, state, LocalRateLimiter, RateLimited, RateLimiter, RawState};
+
+/// A component of a rate limiter.
+///
+/// Each rate limiter may have multiple components. For example, if we are
+/// limiting transaction settlement rate and peanut rate, there will be two type
+/// tags corresponding to the two components, i.e. [`component::SendTx`] and
+/// (hypothetical) `component::PeanutRate`, each implementing [Component].
+pub trait Component {
+    /// Time measure for this limiter component.
+    type Instant: Copy + Ord;
+
+    /// Information in case of rate limit being reached.
+    type LimitedInfo;
+
+    /// Internal state type for this component.
+    type State: RawState<Instant = Self::Instant, LimitedInfo = Self::LimitedInfo>;
+
+    /// Check various preconditions before the rate limiter proper is applied.
+    fn precondition(_limiter: &mut LocalRateLimiter) -> Result<(), RateLimited> {
+        Ok(())
+    }
+
+    /// Get the rate limiter for this component.
+    fn component(limiter: &mut LocalRateLimiter) -> &mut RateLimiter<Self::State>;
+
+    /// Convert the component-specific error into the rate limited error.
+    fn error(info: limiter::RateLimited<Self::LimitedInfo>) -> RateLimited;
+}
+
+/// The rate limiter component for `sendTx` settlement.
+pub enum SendTx {}
```

Comment:
Why is it an `enum`?

Reply:
> Why is it an `enum`?

This is used purely as a phantom type tag. Since it makes no sense to construct an object of this type, I opted for something that cannot be constructed to begin with (there are no enum arms).